# Chapter 4: Control Flow – If-Else, Match & Loops

**Goal**: Learn how to make decisions and repeat tasks – the core of any analytical logic.

---

## 5.1 If – Elif – Else (Decision Making)

Business rules are full of conditions:  
- If sales > target → bonus  
- If margin < 15% → review pricing  
- If customer is Premium → special offer

In [1]:
sales = 1350000
target = 1200000

if sales > target:
    print("Target Achieved! Congratulations.")
elif sales == target:
    print("Exactly met the target.")
else:
    print("Target missed. Action required.")

Target Achieved! Congratulations.


In [2]:
# Nested conditions & business grading
score = 87

if score >= 90:
    grade = "Outstanding"
elif score >= 75:
    grade = "Excellent"
elif score >= 60:
    grade = "Good"
elif score >= 40:
    grade = "Average"
else:
    grade = "Needs Improvement"

print(f"Performance Grade: {grade}")

Performance Grade: Excellent


In [3]:
# Multiple conditions
revenue = 1850000
margin = 0.18
region = "West"

if revenue > 1500000 and margin > 0.15:
    print("High performing region – eligible for expansion budget")
elif revenue > 1000000 or margin > 0.20:
    print("Solid performance – maintain current strategy")
else:
    print("Underperforming – deep dive required")

High performing region – eligible for expansion budget


---

## 5.2 Match-Case (Python 3.10+) – Cleaner Multi-way Branching

Useful when you have many discrete categories (regions, product lines, customer tiers).

In [4]:
region = "West"

match region:
    case "North":
        manager = "Anita Sharma"
        target = 1500000
    case "South":
        manager = "Ravi Kumar"
        target = 1200000
    case "East":
        manager = "Priya Patel"
        target = 1100000
    case "West":
        manager = "Vikram Singh"
        target = 1600000
    case _:
        manager = "Unassigned"
        target = 0

print(f"Region: {region} | Manager: {manager} | Target: ₹{target:,}")

Region: West | Manager: Vikram Singh | Target: ₹1,600,000


---

## 5.3 While Loops – Repeat Until Condition is False

Less common in data analysis than `for` loops, but useful for simulations or waiting for a condition.

In [5]:
# Simple while example – compound interest simulation
principal = 100000
rate = 0.08
years = 0
target = 200000

while principal < target:
    principal *= (1 + rate)
    years += 1

print(f"It will take {years} years to double the money at 8% interest.")
print(f"Final amount: ₹{principal:,.0f}")

It will take 10 years to double the money at 8% interest.
Final amount: ₹215,892


---

## 5.4 For Loops – The Workhorse of Data Analysis

You will use `for` loops constantly until you switch fully to Pandas vectorized operations.

In [6]:
regions = ["North", "South", "East", "West"]
sales = [1250000, 980000, 1120000, 1450000]

for region in regions:
    print(f"Processing {region} region...")

Processing North region...
Processing South region...
Processing East region...
Processing West region...


In [7]:
# Looping with index
for i in range(len(regions)):
    print(f"{regions[i]:8} → ₹{sales[i]:,}")

North    → ₹1,250,000
South    → ₹980,000
East     → ₹1,120,000
West     → ₹1,450,000


In [8]:
# zip() – elegant way to loop over multiple lists together
for region, amount in zip(regions, sales):
    print(f"{region:8} → ₹{amount:,}")

North    → ₹1,250,000
South    → ₹980,000
East     → ₹1,120,000
West     → ₹1,450,000


In [9]:
# range() function
print(list(range(5)))          # 0 to 4
print(list(range(1, 6)))       # 1 to 5
print(list(range(0, 11, 2)))   # Even numbers

[0, 1, 2, 3, 4]
[1, 2, 3, 4, 5]
[0, 2, 4, 6, 8, 10]


In [10]:
# Looping over dictionary
region_sales = {"North": 1250000, "South": 980000, "East": 1120000, "West": 1450000}

for region, amount in region_sales.items():
    status = "Above Target" if amount > 1100000 else "Below Target"
    print(f"{region:8} | ₹{amount:,} | {status}")

North    | ₹1,250,000 | Above Target
South    | ₹980,000 | Below Target
East     | ₹1,120,000 | Above Target
West     | ₹1,450,000 | Above Target


---

## 5.5 Break & Continue

In [11]:
# break – exit the loop early
sales = [850, 920, 1100, 780, 1350, 990]

for s in sales:
    if s > 1200:
        print(f"Found high performer: {s}. Stopping search.")
        break
    print(f"Checking {s}...")

Checking 850...
Checking 920...
Checking 1100...
Checking 780...
Found high performer: 1350. Stopping search.


In [12]:
# continue – skip current iteration
for s in sales:
    if s < 900:
        continue          # Skip low values
    print(f"Good sales: {s}")

Good sales: 920
Good sales: 1100
Good sales: 1350
Good sales: 990


---

## 5.6 Business Mini-Case: Commission Calculator

In [13]:
sales_team = {
    "Anita": 1450000,
    "Ravi": 980000,
    "Priya": 1620000,
    "Vikram": 875000,
    "Sneha": 2100000
}

print("SALES COMMISSION REPORT")
print("=" * 50)

for name, amount in sales_team.items():
    if amount >= 2000000:
        rate = 0.08
    elif amount >= 1500000:
        rate = 0.06
    elif amount >= 1000000:
        rate = 0.04
    else:
        rate = 0.02
    
    commission = amount * rate
    print(f"{name:10} | Sales: ₹{amount:>10,} | Rate: {rate*100:.0f}% | Commission: ₹{commission:>9,.0f}")

print("=" * 50)

SALES COMMISSION REPORT
Anita      | Sales: ₹ 1,450,000 | Rate: 4% | Commission: ₹   58,000
Ravi       | Sales: ₹   980,000 | Rate: 2% | Commission: ₹   19,600
Priya      | Sales: ₹ 1,620,000 | Rate: 6% | Commission: ₹   97,200
Vikram     | Sales: ₹   875,000 | Rate: 2% | Commission: ₹   17,500
Sneha      | Sales: ₹ 2,100,000 | Rate: 8% | Commission: ₹  168,000


---

## Business Challenge – Chapter 4

**Scenario**: You have monthly sales for a product line.

```python
monthly = [185, 210, 195, 240, 175, 260, 230, 290, 215, 250, 280, 310]  # in lakhs
```

**Tasks**:
1. Calculate total annual sales.
2. Find how many months exceeded 250 lakhs.
3. Classify each month as "High" (>250), "Medium" (200-250), or "Low" (<200).
4. Print a month-by-month classification report (use month numbers 1-12).
5. Identify the first month that crossed 300 lakhs (use break).

In [14]:
# Your solution here


---

## Key Takeaways
- `if-elif-else` for decision logic
- `match-case` for clean multi-category branching
- `for` loops + `zip` + dictionaries = powerful data processing
- `break` and `continue` give fine control

**Next Chapter**: Functions & Modules – Writing Reusable Business Logic